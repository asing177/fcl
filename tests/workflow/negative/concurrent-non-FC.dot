digraph workflow {
edge [color="#000000"]
graph [bgcolor=transparent]


initial [shape=point; width=0.3; style=filled; color="#0e64ce"; label=""]
doneAlice [shape=ellipse; fontname="Arial"; fontsize=16; style=filled; color=black; fillcolor=white;]
doneBob [shape=ellipse; fontname="Arial"; fontsize=16; style=filled; color=black; fillcolor=white;]
todoAlice [shape=ellipse; fontname="Arial"; fontsize=16; style=filled; color=black; fillcolor=white;]
todoBob [shape=ellipse; fontname="Arial"; fontsize=16; style=filled; color=black; fillcolor=white;]
terminal [shape=point; width=0.3; peripheries=2; style=filled; color="#d11010"; label=""]

1[label=<<FONT POINT-SIZE="16">requestValues</FONT><FONT POINT-SIZE="10" COLOR="blue"> </FONT>>shape=box; fontname="Arial"; style=filled; color=black; fillcolor=gray75;]
2[label=<<FONT POINT-SIZE="16">setValueAlice</FONT><FONT POINT-SIZE="10" COLOR="blue"> <BR/>roles: alice</FONT>>shape=box; fontname="Arial"; style=filled; color=black; fillcolor=gray75;]
3[label=<<FONT POINT-SIZE="16">setValueBob</FONT><FONT POINT-SIZE="10" COLOR="blue"> <BR/>roles: bob</FONT>>shape=box; fontname="Arial"; style=filled; color=black; fillcolor=gray75;]
4[label=<<FONT POINT-SIZE="16">calculateTotal</FONT><FONT POINT-SIZE="10" COLOR="blue"> </FONT>>shape=box; fontname="Arial"; style=filled; color=black; fillcolor=gray75;]
5[label=<<FONT POINT-SIZE="16">bad</FONT><FONT POINT-SIZE="10" COLOR="blue"> </FONT>>shape=box; fontname="Arial"; style=filled; color=black; fillcolor=gray75;]

initial -> 1 ; 1 -> {todoAlice, todoBob}
todoAlice -> 2 ; 2 -> doneAlice
todoBob -> 3 ; 3 -> doneBob
{doneAlice, doneBob} -> 4 ; 4 -> terminal
doneBob -> 5 ; 5 -> terminal

{rank=same doneAlice doneBob}
{rank=same todoAlice todoBob}


}
