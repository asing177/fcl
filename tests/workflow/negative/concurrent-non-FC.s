global account alice;
global account bob;
global int valueAlice;
global int valueBob;
global int total;

@initial
requestValues(account a, account b) {
  alice = a;
  bob = b;
  transitionTo(@{todoAlice, todoBob});
}

@todoAlice [role: alice]
setValueAlice(int val) {
  valueAlice = val;
  transitionTo(@doneAlice);
}

@todoBob [role: bob]
setValueBob(int val) {
  valueBob = val;
  transitionTo(@doneBob);
}

@{doneAlice, doneBob}
calculateTotal() {
  total = valueAlice + valueBob;
  terminate();
}

@{doneBob}
bad() {
  terminate();
}
