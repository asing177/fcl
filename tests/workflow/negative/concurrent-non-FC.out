Workflow soundness errors:
• Improper completion. Applying 'transition {doneBob} -> {terminal}' to state '{doneAlice
                                                                               ,doneBob}' would result in the new state '{doneAlice
                                                                                                                         ,terminal}' where the terminal place is reached while other places are still active.
• Improper completion. Applying 'transition {doneBob} -> {terminal}' to state '{doneBob
                                                                               ,todoAlice}' would result in the new state '{todoAlice
                                                                                                                           ,terminal}' where the terminal place is reached while other places are still active.